dags/tpu_observability/utils/node_pool_util.py

```diff
@@ -0,0 +1,308 @@
+def _query_status_metric(node_pool: Info) -> Status:
+  """Queries the latest status of the specified GKE node pool.
+  This function retrieves the status by querying the metric
+  "kubernetes.io/node_pool/status" via the Google Cloud Monitoring API.
+
+  Args:
+      node_pool: An instance of the Info class that encapsulates
+                   the configuration and metadata of a GKE node pool.
+
+  Returns:
+      A Status Enum object representing the latest status of the node pool.
+  """
+  monitoring_client = monitoring_v3.MetricServiceClient()
+  project_name = f"projects/{node_pool.project_id}"
+  now = int(time.time())
+  request = monitoring_v3.ListTimeSeriesRequest(
+      name=project_name,
+      filter=(
+          'metric.type="kubernetes.io/node_pool/status" '
+          f'resource.labels.project_id = "{node_pool.project_id}" '
+          f'resource.labels.cluster_name = "{node_pool.cluster_name}" '
+          f'resource.labels.node_pool_name = "{node_pool.node_pool_name}"'
+      ),
+      interval=types.TimeInterval(
+          {
+              "end_time": {"seconds": now},
+              # Metrics are sampled every 60s and stored in the GCP backend,
+              # but it may take up to 2 minutes for the data to become available on the
+              # client side.
+              # Therefore, a longer time interval is necessary.
+              # A 5-minute window is an arbitrary but sufficient choice to ensure we
+              # can retrieve the latest metric data.
+              "start_time": {"seconds": now - 300},
+          }
+      ),
+      view="FULL",
+  )
+
+  # A single query to the Monitoring API can return multiple TimeSeries objects,
+  # especially if the 'status' label changed within the time window (e.g., from
+  # 'PROVISIONING' to 'RUNNING').
+  #
+  # To robustly find the absolute latest status, this block first aggregates all
+  # data points from all series into a single flat list ('records'). It then
+  # finds the record with the maximum timestamp from this list to ensure the
+  # true latest status is identified.
+  time_series_data = monitoring_client.list_time_series(request)
+  records = []
+  for series in time_series_data:
+    np_status = series.metric.labels.get("status", "unknown").upper()
+    for point in series.points:
+      end_ts_dt = point.interval.end_time
+      records.append((end_ts_dt, np_status))
+  if not records:
+    return Status.UNKNOWN
+
+  _, latest_status = max(records, key=lambda r: r[0])
+
+  return Status.from_str(latest_status)
+
+
+@task.sensor(poke_interval=60, timeout=600, mode="reschedule")
+def wait_for_status(
+    node_pool: Info,
+    status: Status,
+    **context,
+) -> bool:
+  """Waits for the node pool to enter the target status.
+
+  This is a task waits for the node pool to enter the target status by querying
+  the status metric and comparing it with the expected status.
+  The default poke_interval and timeout, and document that caller can override them
```

Comment:
this is also an instruction to you

to avoid future confusions:
if I ever use a code-block, you can use them directly, 
but if something that is not wrapped  in a code-block, that could be an instruction or some explanations

---

for your reference, this is a code block
```
codes
```

and this is a suggestion block, which let's just say it's also a code block
```suggestion
  The default poke_interval and timeout, and document that caller can override them.
```

and this is my original comment (quote as below), which it's not a code block
> let's also mention
>
> 1. this is a task waits xxxx by retry xxxxx (action/something)
> 2. the default poke_interval and timeout, and document that caller can override them

Reply:
Got it. I will notice that.